### PR TITLE
LangChain Quickstart

### PR DESCRIPTION
**Pull request by sjyothi54**
**Ticket URL:** https://issues.newrelic.com/browse/NR-156585
**Description:** Created LangChain quickstart

## Pre merge checklist

- [ ] Did you check you NRQL syntax? - Does it work?
- [ ] Did you include a Data source and Documentation reference?
- [ ] Are all documentation links publicly accessible?
- [ ] Did you check your descriptive content for [voice and tone](https://docs.newrelic.com/docs/style-guide/writing-strategies/voice-strategies-docs-sound-new-relic/)?
- [ ] Did you check your descriptive content for spelling and grammar errors?
- [ ] Did you review your content with a subject matter expert? (e.g. a Browser agent quickstart is reviewed with a member of the Browser Agent team)

### Dashboards

- [ ] Does the PR contain a screenshot for each of your dashboards?
- [ ] Do your screenshots show data?
- [ ] Has the [sanitization script](https://github.com/newrelic/newrelic-quickstarts/blob/main/CONTRIBUTING.md#dashboards) been run on each dashboard?

### Alerts

- [ ] Did you check that your alerts actually work?

### CODE REVIEW

File: quickstarts/mlops/langchain/config.yml [hunk 1]
```diff
@@ -0,0 +1,46 @@
+slug: langchain
+description: |
+  ## Why monitor LangChain?
+  LangChain is a framework for developing applications powered by language models. New Relic LangChain quickstart performance monitoring offers immediate and pre-configured observability.
+  With this quickstart, you will:
```
Comment (rahulbasu-nr) [inline quickstarts/mlops/langchain/config.yml:5]:
```suggestion
  With this quickstart, you will be able:
```

File: quickstarts/mlops/langchain/config.yml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+  ### To ensure uptime: 
+  LangChain AI is a complex system, and it is important to make sure that it is always up and running. By monitoring the system, you will quickly identify and fix any issues that may cause downtime.
+  ### To track performance:
```
Comment (rahulbasu-nr) [inline quickstarts/mlops/langchain/config.yml:8]:
```suggestion
  ### Track performance:  
```

File: quickstarts/mlops/langchain/config.yml [hunk 3]
```diff
@@ -0,0 +1,46 @@
+slug: langchain
+description: |
+  ## Why monitor LangChain?
+  LangChain is a framework for developing applications powered by language models. New Relic LangChain quickstart performance monitoring offers immediate and pre-configured observability.
+  With this quickstart, you will:
+  ### To ensure uptime: 
+  LangChain AI is a complex system, and it is important to make sure that it is always up and running. By monitoring the system, you will quickly identify and fix any issues that may cause downtime.
+  ### To track performance:  
+  Monitoring the performance of LangChain AI will help you identify areas where you will improve. For example, you will track things like latency, throughput, and error rates.
+  ### To improve the model:
```
Comment (rahulbasu-nr) [inline quickstarts/mlops/langchain/config.yml:10]:
```suggestion
  ### Improve the model: 
```

File: quickstarts/mlops/langchain/config.yml [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+description: |
+  ## Why monitor LangChain?
+  LangChain is a framework for developing applications powered by language models. New Relic LangChain quickstart performance monitoring offers immediate and pre-configured observability.
+  With this quickstart, you will:
+  ### To ensure uptime: 
+  LangChain AI is a complex system, and it is important to make sure that it is always up and running. By monitoring the system, you will quickly identify and fix any issues that may cause downtime.
+  ### To track performance:  
+  Monitoring the performance of LangChain AI will help you identify areas where you will improve. For example, you will track things like latency, throughput, and error rates.
+  ### To improve the model: 
+  By monitoring the performance of LangChain AI, you will identify areas where the model will be improved. For example, you will track things like the accuracy of the model's predictions and the diversity of the model's responses.
```
Comment (rahulbasu-nr) [inline quickstarts/mlops/langchain/config.yml:11]:
```suggestion
  Monitoring the performance of LangChain AI allows for the identification of areas where the model can be improved. For example, tracking the accuracy of the model's predictions and the diversity of the model's responses.
```

File: quickstarts/mlops/langchain/config.yml [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+  
+  ## Comprehensive monitoring quickstart for LangChain
+  Our LangChain quickstart provides you metrics such as uptime, accuracy, relevance, and speed. These metrics can help you ensure that LangChain AI is performing at its best.
```
Comment (rahulbasu-nr) [inline quickstarts/mlops/langchain/config.yml:14]: Our LangChain quickstart provides metrics, including uptime, accuracy, relevance, and speed. These metrics serve as valuable tools to empower you in maximizing the performance of LangChain AI.

File: quickstarts/mlops/langchain/config.yml [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  ### To track performance:  
+  Monitoring the performance of LangChain AI will help you identify areas where you will improve. For example, you will track things like latency, throughput, and error rates.
+  ### To improve the model: 
+  By monitoring the performance of LangChain AI, you will identify areas where the model will be improved. For example, you will track things like the accuracy of the model's predictions and the diversity of the model's responses.
+  
+  ## Comprehensive monitoring quickstart for LangChain
+  Our LangChain quickstart provides you metrics such as uptime, accuracy, relevance, and speed. These metrics can help you ensure that LangChain AI is performing at its best.  
+  
+  ## What’s included in this quickstart?
+  New Relic LangChain monitoring quickstart provides quality out-of-the-box reporting:
```
Comment (rahulbasu-nr) [inline quickstarts/mlops/langchain/config.yml:17]:
```suggestion
  New Relic LangChain monitoring quickstart offers a range of out-of-the-box reports, including:
```

File: quickstarts/mlops/langchain/config.yml [hunk 7]
```diff
@@ -0,0 +1,51 @@
+slug: langchain
+description: |
+  ## Why monitor LangChain?
+  LangChain is a framework for developing applications powered by language models. New Relic LangChain quickstart performance monitoring offers immediate and pre-configured observability.
+  By using the LangChain quickstart, you will be able to:
+  ### Ensure uptime: 
+  LangChain AI is a complex system, and it is important to make sure that it is always up and running. By monitoring the system, you will quickly identify and fix any issues that may cause downtime.
+  ### Track performance:  
+  Monitoring the performance of LangChain AI will help you identify areas where you will improve. For example, you will track things like latency, throughput, and error rates.
```
Comment (d3caf) [inline quickstarts/mlops/langchain/config.yml:9]: Maybe rephrase these sentences?

File: quickstarts/mlops/langchain/config.yml [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+  
+  ## Comprehensive monitoring quickstart for LangChain
+  Our LangChain quickstart provides metrics, including uptime, accuracy, relevance, and speed. These metrics serve as valuable tools to empower you in maximizing the performance of LangChain AI.
+  
+  ## What’s included in this quickstart?
+  New Relic LangChain monitoring quickstart offers a range of out-of-the-box reports, including:
+  - Gain code-related insights about your LangChain's health and status
+  - Monitor your LangChain latency and tokens
+  - Entity (average tokens, cost, total requests, api response time and many more)
+  - Alerts (errors,requests per model and response time)
```
Comment (d3caf) [inline quickstarts/mlops/langchain/config.yml:21]: Reword this a bit as well. The `including` doesn't really flow with the bullet points

**Review by d3caf — APPROVED** [review]

File: quickstarts/mlops/langchain/config.yml (hunk 7 above)
Comment (rahulbasu-nr) [inline quickstarts/mlops/langchain/config.yml:9]:
```suggestion
  By tracking key metrics like latency, throughput, and error rates, you can gain insights into LangChain AI's performance and identify areas of improvement.
```

File: quickstarts/mlops/langchain/config.yml [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+  ### Track performance:  
+  Monitoring the performance of LangChain AI will help you identify areas where you will improve. For example, you will track things like latency, throughput, and error rates.
+  ### Improve the model: 
+  Monitoring the performance of LangChain AI allows for the identification of areas where the model can be improved. For example, tracking the accuracy of the model's predictions and the diversity of the model's responses.
+  
+  ## Comprehensive monitoring quickstart for LangChain
+  Our LangChain quickstart provides metrics, including uptime, accuracy, relevance, and speed. These metrics serve as valuable tools to empower you in maximizing the performance of LangChain AI.
+  
+  ## What’s included in this quickstart?
+  New Relic LangChain monitoring quickstart offers a range of out-of-the-box reports, including:
```
Comment (rahulbasu-nr) [inline quickstarts/mlops/langchain/config.yml:17]:
```suggestion
 New Relic LangChain monitoring quickstart provides a variety of pre-built reports, which will help you gain insights into the health and performance of your LangChain. These reports include:
```

File: quickstarts/mlops/langchain/config.yml [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+  - Gain code-related insights about your LangChain's health and status
```
Comment (rahulbasu-nr) [inline quickstarts/mlops/langchain/config.yml:18]:
```suggestion
  - Get code-related insights to identify potential problems with your code and improve its performance.
```

File: quickstarts/mlops/langchain/config.yml [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+  - Monitor your LangChain latency and tokens
```
Comment (rahulbasu-nr) [inline quickstarts/mlops/langchain/config.yml:19]: Track the latency details and number of tokens

File: quickstarts/mlops/langchain/config.yml [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+  - Entity (average tokens, cost, total requests, api response time and many more)
```
Comment (rahulbasu-nr) [inline quickstarts/mlops/langchain/config.yml:20]:
```suggestion
  - Entity details such as the average number of tokens used, the cost per entity, and the total number of requests made
```

File: quickstarts/mlops/langchain/config.yml (hunk 8 above)
Comment (rahulbasu-nr) [inline quickstarts/mlops/langchain/config.yml:21]:
```suggestion
  - Alerts for errors, requests per model, and response time.
```

File: quickstarts/mlops/langchain/config.yml [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+  - Alerts (errors,requests per model and response time)
+summary: |
+  Use the New Relic LangChain quickstart to improve the performance of your LangChain.
```
Comment (rahulbasu-nr) [inline quickstarts/mlops/langchain/config.yml:23]:
```suggestion
  Improve your LangChain Performance with New Relic LangChain Quickstart.
```

File: alert-policies/langchain/error.yml [hunk 14]
```diff
@@ -0,0 +1,27 @@
+name: Errors
+
+description: |+
+  This alert is triggered if the number of errors is greater than 0 within 5 minutes.
```
Comment (rahulbasu-nr) [inline alert-policies/langchain/error.yml:4]:
```suggestion
  This alert is triggered if the number of errors is greater than 0 for 5 minutes.
```

File: alert-policies/langchain/requestsPerModel.yml [hunk 15]
```diff
@@ -0,0 +1,35 @@
+name: Requests per model
+
+description: |+
+  This alert is triggered if requests per model exceeds 1000 in 5 minutes.
```
Comment (rahulbasu-nr) [inline alert-policies/langchain/requestsPerModel.yml:4]:
```suggestion
  This alert is triggered if requests per model exceeds 1000 for 5 minutes.
```

File: alert-policies/langchain/responseTime.yml [hunk 16]
```diff
@@ -0,0 +1,35 @@
+name: Response time
+
+description: |+
+  This alert is triggered if response time exceeds 5 seconds in 5 minutes.
```
Comment (rahulbasu-nr) [inline alert-policies/langchain/responseTime.yml:4]:
```suggestion
  This alert is triggered if response time exceeds 5 seconds for 5 minutes.
```

File: quickstarts/mlops/langchain/config.yml [hunk 17]
```diff
@@ -0,0 +1,51 @@
+slug: langchain
+description: |
+  ## Why monitor LangChain?
+  LangChain is a framework for developing applications powered by language models. New Relic LangChain quickstart performance monitoring offers immediate and pre-configured observability.
```
Comment (rahulbasu-nr) [inline quickstarts/mlops/langchain/config.yml:4]:
```suggestion
 Monitoring LangChain will help you to ensure its performance, identify and prevent outages, detect security threats, gather performance metrics, and identify potential opportunities for collaboration or integration.
```

File: quickstarts/mlops/langchain/config.yml [hunk 18]
```diff
@@ -0,0 +1,51 @@
+slug: langchain
+description: |
+  ## Why monitor LangChain?
```
Comment (rahulbasu-nr) [inline quickstarts/mlops/langchain/config.yml:3]:
```suggestion
  ## Why monitor your LangChain powered application?
```

File: quickstarts/mlops/langchain/config.yml (hunk 17 above)
Comment (rahulbasu-nr) [inline quickstarts/mlops/langchain/config.yml:4]:
```suggestion
  LangChain is a framework for developing applications powered by language models. By monitoring your LangChain app, we are providing you visibility on the chains, tools as well as inputs and outputs.
```

File: quickstarts/mlops/langchain/config.yml [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+  LangChain is a framework for developing applications powered by language models. New Relic LangChain quickstart performance monitoring offers immediate and pre-configured observability.
+  By using the LangChain quickstart, you will be able to:
+  ### Ensure uptime:
```
Comment (rahulbasu-nr) [inline quickstarts/mlops/langchain/config.yml:6]:
```suggestion
   
```

File: quickstarts/mlops/langchain/config.yml [hunk 20]
```diff
@@ -0,0 +1,51 @@
+slug: langchain
+description: |
+  ## Why monitor LangChain?
+  LangChain is a framework for developing applications powered by language models. New Relic LangChain quickstart performance monitoring offers immediate and pre-configured observability.
+  By using the LangChain quickstart, you will be able to:
+  ### Ensure uptime: 
+  LangChain AI is a complex system, and it is important to make sure that it is always up and running. By monitoring the system, you will quickly identify and fix any issues that may cause downtime.
```
Comment (rahulbasu-nr) [inline quickstarts/mlops/langchain/config.yml:7]:
```suggestion
```

Comment (rahulbasu-nr) [inline quickstarts/mlops/langchain/config.yml:7]:
```suggestion
  
```

File: quickstarts/mlops/langchain/config.yml [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+  ### Track performance:  
+  By tracking key metrics like latency, throughput, and error rates, you can gain insights into LangChain AI's performance and identify areas of improvement.
```
Comment (rahulbasu-nr) [inline quickstarts/mlops/langchain/config.yml:9]:
```suggestion
  ### Track your app:  
  By tracking key metrics like latency, throughput, error rates, and input & output, you can gain insights into your LangChain app's performance and identify areas of improvement.
```

File: quickstarts/mlops/langchain/config.yml [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+  ### Improve the model: 
+  Monitoring the performance of LangChain AI allows for the identification of areas where the model can be improved. For example, tracking the accuracy of the model's predictions and the diversity of the model's responses.
```
Comment (rahulbasu-nr) [inline quickstarts/mlops/langchain/config.yml:11]:
```suggestion
  ### Track the LLM's performance: 
  Monitor the input/output, latency and errors of your LLM provider. Track performance changes depending upon the versions and providers of your LLM. 
  
  
  Monitor the usage to understand the cost, rate limits and general performance.
  
  
  
```

File: quickstarts/mlops/langchain/config.yml [hunk 23]
```diff
@@ -0,0 +1,51 @@
...
+  
+  ## Comprehensive monitoring quickstart for LangChain
+  Our LangChain quickstart provides metrics, including uptime, accuracy, relevance, and speed. These metrics serve as valuable tools to empower you in maximizing the performance of LangChain AI.
```
Comment (rahulbasu-nr) [inline quickstarts/mlops/langchain/config.yml:14]:
```suggestion
  ## Comprehensive monitoring quickstart for your LangChain app
  Our LangChain quickstart provides metrics, including error rate, input/output, latency, queries, retrieved document content, integrate different LLM providers and vector stores.
```

File: quickstarts/mlops/langchain/config.yml [hunk 24]
```diff
@@ -0,0 +1,51 @@
...
+  
+  ## What’s included in this quickstart?
+  New Relic LangChain monitoring quickstart provides a variety of pre-built reports, which will help you gain insights into the health and performance of your LangChain. These reports include:
```
Comment (rahulbasu-nr) [inline quickstarts/mlops/langchain/config.yml:17]:
```suggestion
  New Relic LangChain monitoring quickstart provides a variety of pre-built reports, which will help you gain insights into the health and performance of your LangChain app. These reports include:
```

File: quickstarts/mlops/langchain/config.yml [hunk 25]
```diff
@@ -0,0 +1,51 @@
...
+  - Code-related insights to identify potential problems with your code and improve its performance
+  - Latency details and number of tokens
+  - Entity details such as the average number of tokens used, the cost per entity, and the total number of requests made
+  - Alerts for errors,requests per model and response time
```
Comment (rahulbasu-nr) [inline quickstarts/mlops/langchain/config.yml:21]:
```suggestion
  - LLM calls and chains
  - Tools
  - Vector searches
  - Alerts for errors, requests per model and response time
```

File: quickstarts/mlops/langchain/config.yml [hunk 26]
```diff
@@ -0,0 +1,51 @@
...
+level: 
+  - New Relic
+authors:
+  - New Relic
+  - Jyothi Surampudi
+title: LangChain
+documentation:
+  - name: LangChain integration documentation
+    description: |
+     Implement monitoring and instrumentation for your LangChain, and ensure that your LangChain data is integrated into New Relic for effective performance analysis and insights.
```
Comment (rahulbasu-nr) [inline quickstarts/mlops/langchain/config.yml:34]:
```suggestion
     Implement monitoring and instrumentation for your LangChain app, to ensure that your observability data is integrated into New Relic for effective performance analysis and insights.
```

File: alert-policies/langchain/error.yml (hunk 14 above)
Comment (rahulbasu-nr) [inline alert-policies/langchain/error.yml:4]:
```suggestion
  This alert is triggered if the number of errors exceeds 2 within 5 minutes.
```

File: alert-policies/langchain/error.yml [hunk 27]
```diff
@@ -0,0 +1,27 @@
...
+# Function used to aggregate the NRQL query value(s) for comparison to the terms.threshold (Default: SINGLE_VALUE)
+valueFunction: SINGLE_VALUE
+
+# List of Critical and Warning thresholds for the condition
+terms:
+  - priority: CRITICAL
+    # Operator used to compare against the threshold.
+    operator: ABOVE
+    # Value that triggers a violation
+    threshold: 0
```
Comment (rahulbasu-nr) [inline alert-policies/langchain/error.yml:18]:
```suggestion
    threshold: 2
```

**Review by rahulbasu-nr — CHANGES_REQUESTED** [review]
.

File: data-sources/langchain/config.yml [hunk 28]
```diff
@@ -0,0 +1,20 @@
+id: langchain
+displayName: LangChain
+description: |
+  Implement monitoring and instrumentation for your LangChain, and ensure that your LangChain data is integrated into New Relic for effective performance analysis and insights.
+install:
+  primary:
+    link:
+      url: https://github.com/newrelic/nr-openai-observability
```
Comment (pkosanam) [inline data-sources/langchain/config.yml:8]:
```suggestion
    nerdlet:
      nerdletId: marketplace.install-data-source
      nerdletState:
        dataSourceId: langchain-vectordb
      requiresAccount: false
```

File: data-sources/langchain/config.yml [hunk 29]
```diff
@@ -0,0 +1,20 @@
+id: langchain
+displayName: LangChain
+description: |
+  Implement monitoring and instrumentation for your LangChain, and ensure that your LangChain data is integrated into New Relic for effective performance analysis and insights.
+install:
+  primary:
+    link:
```
Comment (pkosanam) [inline data-sources/langchain/config.yml:7]:
```suggestion
```

**Review by pkosanam — CHANGES_REQUESTED** [review]

**Review by mdumpati — APPROVED** [review]
LGTM

**Review by pkosanam — DISMISSED** [review]
LGTM

**Review by seemantk — DISMISSED** [review]
lgtm

File: dashboards/langchain/langchain.json [hunk 30]
```diff
@@ -0,0 +1,792 @@
+{
+  "name": "LangChain",
```
Comment (pkosanam) [inline dashboards/langchain/langchain.json:2]:
```suggestion
  "name": "LangChain Application",
```

**Review by mdumpati — APPROVED** [review]
LGTM

**Review by seemantk — APPROVED** [review]
lgtm
